### ISSUE TITLE
Monadic binding operators don't allow attributes

### ISSUE THREAD

**Issue by anmonteiro** [ocaml/ocaml#9301]
I tend to use `let[@ocaml.warning "-8"] ...` when I'm trying something out quickly.

This is not currently possible with binding operators, and instead produces a syntax error:

```ocaml
File "src/redacted.ml", line 63, characters 8-10:
63 |   let* [@ocaml.warning "-8"] (Some v) =
             ^^
Error: Syntax error
```

**Comment by anmonteiro** [comment]
still relevant.

**Comment by gasche** [comment]
@Octachron I marked this "good first issue" as it sounds like something not-too-hard for someone willing to get introduced to the parser stuff. Did you self-assign as you are planning to work on this shortly?

**Comment by Octachron** [comment]
It is a good first issue indeed. I was more thinking to add it to my medium term stack of small fixes for 5.1. I am thus not planning to work on this shortly, and I will gladly review a PR fixing the issue.
Like this we have both a `good first issue` label and pre-assigned reviewer.

**Comment by maurobringolf** [comment]
I also ran into this and plan to start working on a PR. Any guidance is welcome, this is my first exposure to the codebase.

**Comment by gasche** [comment]
parsing/parsetree.mli has the AST definition, with documentation -- the relevant nodes are reachable by looking for `letop`. Then parsing/parser.mly has the parser itself, looking for `letop` there will point at the place to modify. Look at `%inline let_binding(EXT):` and the `mklb` helper for inspiration on how attributes are handled for the usual `let` rule. You may need to modify the AST definition for binding operators (in parsetree.mli) to add attribute information.

See parsing/HACKING.adoc for documentation on how to "compile" parser.mly.

To test the change, I would add uses of attributes on binding operators in `testsuite/tests/parsetree/source.ml` (there are already binding operators in the file). The test (`./ocamltest/ocamltest testsuite/tests/parsetree/test.ml`) checks that parsing this file, printing it and parsing the result gives back the same AST. This means that you may have to fix the source printer in parsing/pprintast.ml (this is used for the output of `ocamlc -dsource foo.ml`); while you are at it, parsing/printast.ml would probably also need fixing (this is used for the output of `ocamlc -dparsetree foo.ml`).

**Comment by shindere** [comment]
I find your explanations very cool and complete, @gashce! Thanks!

It really gives one the motivation to start working on it.

I don't want t to take anybody's place, especially given that I am not
at all familiar with monadic binding operators and binders, but in <case
no body could work on it, I can very well see myself volunteering to do
so, if asked.

**Comment by maurobringolf** [comment]
Thanks for the hints @gasche :pray: , I am working on it.

**Comment by maurobringolf** [comment]
> You may need to modify the AST definition for binding operators (in parsetree.mli) to add attribute information.

I think this is necessary to support attributes for each binding separately in a `let* ... and* ... and* ...` block. This first commit I made simply adds attributes to the `letop` node. Since let-bindings support separate attributes for `and`, I think it makes sense to do the same for `and*`.

Do you agree? Should I simply make a proposal (happy to) or do you have a specific AST change in mind? Since this is quite a fundamental thing.

**Comment by gasche** [comment]
I agree that one should be able to place attributes in the same positions as with a normal `let .. and .. ` binding.

**Comment by Oluwabukolab** [comment]
Hello, I am an outreachy applicant, I am interested in working on this issue. Are there any tips you have to get started on this?

**Comment by Oluwabukolab** [comment]
Hello @Octachron, can I work on this issue? Are there any tips to get started? Thank you!

**Comment by gasche** [comment]
@Bukolab99 there is already work in progress on this issue ( #11398 ) so I would recommend looking at something else.

**Comment by Oluwabukolab** [comment]
Alright. Thank you!
